File: cli/src/fp_cli/project.py
"""Project manifest (.fp.yaml) read/write."""

from pathlib import Path

import yaml

PROJECT_FILE = ".fp.yaml"

ALLOWED_EXTENSIONS = {".py", ".css", ".js", ".svg", ".html", ".json", ".txt"}


def find_project_file(start: Path | None = None) -> Path | None:
    """Find .fp.yaml in current or parent directories."""
    current = start or Path.cwd()
    for directory in [current, *current.parents]:
        candidate = directory / PROJECT_FILE
        if candidate.exists():
            return candidate
    return None


def read_project(path: Path | None = None) -> dict:
    """Read and return project config. Exits if not found."""
    fp_file = path or find_project_file()
    if not fp_file or not fp_file.exists():
        from .console import err_console

        err_console.print(
            f"[red]No {PROJECT_FILE} found.[/red] Run [bold]fp init[/bold] first."
        )
        raise SystemExit(1)
    return yaml.safe_load(fp_file.read_text()) or {}


def write_project(data: dict, directory: Path | None = None) -> Path:
    """Write project config to .fp.yaml. Returns the file path."""
    target = (directory or Path.cwd()) / PROJECT_FILE
    target.write_text(yaml.dump(data, default_flow_style=False, sort_keys=False))
    return target


def collect_files(directory: Path | None = None, entrypoint: str = "app.py") -> dict[str, str]:
    """Collect all deployable files from directory."""
    root = directory or Path.cwd()
    files = {}

    for path in sorted(root.rglob("*")):
        if not path.is_file():
            continue
        if path.suffix not in ALLOWED_EXTENSIONS:
            continue

        rel = str(path.relative_to(root))

        # Skip hidden files, __pycache__, .venv, node_modules, etc.
        parts = Path(rel).parts
        if any(p.startswith(".") or p == "__pycache__" or p == ".venv" or p == "node_modules" for p in parts):
            continue

        files[rel] = path.read_text()

    if entrypoint not in files:
        from .console import err_console

        err_console.print(f"[red]Entrypoint '{entrypoint}' not found in {root}[/red]")
        raise SystemExit(1)

    return files


def detect_mode(files: dict[str, str]) -> str:
    """Detect single vs multi mode from collected files."""
    py_files = [f for f in files if f.endswith(".py")]
    return "multi" if len(py_files) > 1 or any(f for f in files if not f.endswith(".py")) else "single"

File: cli-go/internal/project/project.go
package project

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

const (
	ProjectFile = ".fp.yaml"
)

var AllowedExtensions = map[string]bool{
	".py": true, ".css": true, ".js": true, ".svg": true,
	".html": true, ".json": true, ".txt": true,
}

var SkipDirs = map[string]bool{
	"__pycache__": true, ".venv": true, "node_modules": true,
}

type Project struct {
	Name       string            `yaml:"name"`
	Entrypoint string            `yaml:"entrypoint"`
	Env        map[string]string `yaml:"env"`
	Database   interface{}       `yaml:"database"`
}

func FindProjectFile(start string) (string, bool) {
	if start == "" {
		start, _ = os.Getwd()
	}
	dir := start
	for {
		candidate := filepath.Join(dir, ProjectFile)
		if _, err := os.Stat(candidate); err == nil {
			return candidate, true
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return "", false
}

func Read(path string) (*Project, error) {
	if path == "" {
		return nil, fmt.Errorf("no %s found. Run 'fp init' first", ProjectFile)
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var p Project
	if err := yaml.Unmarshal(data, &p); err != nil {
		return nil, err
	}
	if p.Entrypoint == "" {
		p.Entrypoint = "app.py"
	}
	return &p, nil
}

func ReadFromCwd() (*Project, string, error) {
	fp, ok := FindProjectFile("")
	if !ok {
		return nil, "", fmt.Errorf("no %s found. Run 'fp init' first", ProjectFile)
	}
	p, err := Read(fp)
	if err != nil {
		return nil, "", err
	}
	dir := filepath.Dir(fp)
	return p, dir, nil
}

func Write(dir string, p *Project) error {
	data, err := yaml.Marshal(p)
	if err != nil {
		return err
	}
	path := filepath.Join(dir, ProjectFile)
	return os.WriteFile(path, data, 0644)
}

func shouldSkip(rel string) bool {
	parts := strings.Split(filepath.ToSlash(rel), "/")
	for _, p := range parts {
		if strings.HasPrefix(p, ".") || SkipDirs[p] {
			return true
		}
	}
	return false
}

func CollectFiles(dir, entrypoint string) (map[string]string, error) {
	files := make(map[string]string)
	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(dir, path)
		if err != nil {
			return err
		}
		rel = filepath.ToSlash(rel)
		ext := strings.ToLower(filepath.Ext(path))
		if !AllowedExtensions[ext] {
			return nil
		}
		if shouldSkip(rel) {
			return nil
		}
		content, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		files[rel] = string(content)
		return nil
	})
	if err != nil {
		return nil, err
	}
	if _, ok := files[entrypoint]; !ok {
		return nil, fmt.Errorf("entrypoint '%s' not found in %s", entrypoint, dir)
	}
	return files, nil
}

func DetectMode(files map[string]string) string {
	pyCount := 0
	hasNonPy := false
	for f := range files {
		if strings.HasSuffix(f, ".py") {
			pyCount++
		} else {
			hasNonPy = true
		}
	}
	if pyCount > 1 || hasNonPy {
		return "multi"
	}
	return "single"
}
